eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/RelayerSpec.scala

```diff
@@ -0,0 +1,196 @@
+
+    relayer ! RelayForward(add_ab)
+
+    register.expectMessageType[Register.Forward[CMD_FAIL_HTLC]]
+  }
+
+  test("relay an htlc-add at the final node to the payment handler") { f =>
+    import f._
+
+    val (cmd, _) = buildCommand(ActorRef.noSender, Upstream.Local(UUID.randomUUID()), paymentHash, hops.take(1), FinalLegacyPayload(finalAmount, finalExpiry))
+    val add_ab = UpdateAddHtlc(channelId = channelId_ab, id = 123456, cmd.amount, cmd.paymentHash, cmd.cltvExpiry, cmd.onion)
+
+    relayer ! RelayForward(add_ab)
+
+    val fp = paymentHandler.expectMessageType[FinalPacket]
+    assert(fp.add === add_ab)
+    assert(fp.payload === FinalLegacyPayload(finalAmount, finalExpiry))
+
+    register.expectNoMessage(50 millis)
+  }
+
+  test("relay a trampoline htlc-add at the final node to the payment handler") { f =>
+    import PaymentPacketSpec._
+    import f._
+    val a = PaymentPacketSpec.a
+
+    // We simulate a payment split between multiple trampoline routes.
+    val totalAmount = finalAmount * 3
+    val trampolineHops = NodeHop(a, b, channelUpdate_ab.cltvExpiryDelta, 0 msat) :: Nil
+    val (trampolineAmount, trampolineExpiry, trampolineOnion) = OutgoingPacket.buildPacket(Sphinx.TrampolinePacket)(paymentHash, trampolineHops, Onion.createMultiPartPayload(finalAmount, totalAmount, finalExpiry, paymentSecret))
+    assert(trampolineAmount === finalAmount)
+    assert(trampolineExpiry === finalExpiry)
+    val (cmd, _) = buildCommand(ActorRef.noSender, Upstream.Local(UUID.randomUUID()), paymentHash, ChannelHop(a, b, channelUpdate_ab) :: Nil, Onion.createTrampolinePayload(trampolineAmount, trampolineAmount, trampolineExpiry, randomBytes32, trampolineOnion.packet))
+    assert(cmd.amount === finalAmount)
+    assert(cmd.cltvExpiry === finalExpiry)
+    val add_ab = UpdateAddHtlc(channelId = channelId_ab, id = 123456, cmd.amount, cmd.paymentHash, cmd.cltvExpiry, cmd.onion)
+
+    relayer ! RelayForward(add_ab)
+
+    val fp = paymentHandler.expectMessageType[FinalPacket]
+    assert(fp.add === add_ab)
+    assert(fp.payload.amount === finalAmount)
+    assert(fp.payload.totalAmount === totalAmount)
+    assert(fp.payload.expiry === finalExpiry)
+    assert(fp.payload.paymentSecret === Some(paymentSecret))
+
+    register.expectNoMessage(50 millis)
+  }
+
+  test("fail to relay an htlc-add when the onion is malformed") { f =>
+    import f._
+
+    // we use this to build a valid onion
+    val (cmd, _) = buildCommand(ActorRef.noSender, Upstream.Local(UUID.randomUUID()), paymentHash, hops, FinalLegacyPayload(finalAmount, finalExpiry))
+    // and then manually build an htlc with an invalid onion (hmac)
+    val add_ab = UpdateAddHtlc(channelId = channelId_ab, id = 123456, cmd.amount, cmd.paymentHash, cmd.cltvExpiry, cmd.onion.copy(hmac = cmd.onion.hmac.reverse))
+    relayer ! LocalChannelUpdate(null, channelId_bc, channelUpdate_bc.shortChannelId, c, None, channelUpdate_bc, makeCommitments(channelId_bc))
```

Comment:
This line doesn't make sense anymore, remove?

Reply:
Fixed in 626b4027f.